### DISCUSSION TITLE
How can I make Wix 4 treat a lower major version as an upgrade?

### DISCUSSION THREAD

**Discussion by paulvickery** [wixtoolset/issues#7521] [Questions]
Hi,

I have a product which uses the supported year as the major version number, e.g. 2024.3.3544.4567. It has been this way for several years, and only works because I am able to suppress the ICE24 error. Going forward, this suppression may (will) at some point no longer be supported. I know that ideally it would have been better to not have done this, and that I have made trouble for myself, so am looking for a solution.

So, I want to change my versioning to use the 2-digit year, e.g. 24.3.3544.4567, which is what I should have gone for in the first place, but will mess up upgrading previous version installations that use the 4-digit year.

I think from what I have read and heard that it needs to use a custom Upgrade Table, but can't find any real examples of how I would use that for this.

Does anyone know the way to solve this? 

Thanks,

Paul.

**Comment by BMurri** [comment]
First, two items to keep in mind for some background.
1) When you reach the year 2048, internally MSI will replace that with 0 (since it is wrapping it), and all your packages will be considered downgrades (forever more) if you do nothing. That is a big part of why that ICE24 report is an error and not a warning.
2) File versions are not limited to major (and minor) version values being 0-255 (that limit is 0-65535, although warnings start going off past 32767), so if you have been sharing your file versions with your MSI version, you probably will need to break that connection (because all sorts of problems can happen when the file versions aren't always incrementing).

There will be an issue you really cannot solve where already released packages won't be prevented from installing on top of (replacing) your new fixed packages. If you can live with potential side-by-side installs (one pre-fix and one post-fix) and are willing to change your default installation directory, you will have a much easier time moving forward. Otherwise, it will be a bit more complicated, and old packages can still "upgrade" you in ways that will likely leave broken deployments (mix of old and new files, usually).

**Reply by BMurri** [reply]
* there is a way to prevent already released packages from installing over the top of your newer packages, but it involves installing a nearly empty MSI that you will want to guard against removal while your new packages are installed, and doing that is something I don't recommend unless you change to always using a chainer instead of distributing MSI files directly. Let me know if you want to know about this idea. It's a kluge, but it does work.

**Reply by BMurri** [reply]
Now that I've given you a bunch of food for thought, I would very strongly suggest that you consider getting some paid support. You will really want to have some experienced experts in this space evaluate whatever solution you come up with before you go public with it, because there are several considerations that will end up being particular to your own product and general advice made in public will not likely apply to others who paint themselves into this same corner.

If your product is FOSS and you are very budget constrained, let us know, someone may be able to help.
